http_proxy.go

```diff
@@ -826,9 +826,28 @@ func (p *Proxy) listenShadowsocks(addr string) (net.Listener, error) {
 	if err != nil {
 		return nil, errors.New("Unable to create shadowsocks cipher: %v", err)
 	}
-	base, err := net.Listen("tcp", addr)
-	if err != nil {
-		return nil, err
+	var tlsConfig *tls.Config
+	if p.KeyFile != "" && p.CertFile != "" {
+		var cert tls.Certificate
+		cert, err = tls.LoadX509KeyPair(p.CertFile, p.KeyFile)
+		if err != nil {
+			return nil, errors.New("unable to load cert: %v", err)
+		}
+
+		tlsConfig = &tls.Config{Certificates: []tls.Certificate{cert}}
+	}
+
+	var base net.Listener
+	if tlsConfig != nil {
+		base, err = tls.Listen("tcp", addr, tlsConfig)
```

Comment:
This is backwards.  The TLS layer needs to be on top of the Shadowsocks layer.  Using TLS as the transport underneath Shadowsocks, as done here, will result in TLS being visible on the wire, not Shadowsocks.  I think using `tls.NewListener` and wrapping the Shadowsocks listener will give us the desired result.

The ordering of the network stack gets confusing, especially when we do weird stuff like this and compose different application-layer protocols.

Reply:
Aaaaaah got it! Thanks for explaining!